File: 0056-Merge-Intervals/lc0056.rs
// Rust solution.

impl Solution {
    pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut ans: Vec<Vec<i32>> = Vec::new();
        let mut intervals = intervals;
        intervals.sort_by(|a, b| a[0].cmp(&b[0]));
        for (k, v) in intervals.into_iter().enumerate() {
            if k == 0 { ans.push(v); }
            else {
                if let Some(back) = ans.last_mut() {
                    if v[0] > back[1] {
                        ans.push(v);
                    } else {
                        back[1] = std::cmp::max(back[1], v[1]);
                    }
                }
            }
        }
        ans
    }
}

File: 0056-Merge-Intervals/lc0056.cc
// C++ solution.

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> ans;
        sort(intervals.begin(), intervals.end(),
             [](const vector<int>& a, const vector<int>& b) { return a[0] < b[0]; });
        if (intervals.size() > 0) {
            ans.push_back(intervals[0]);
            for (int i = 1; i < intervals.size(); i++) {
                if (intervals[i][0] > ans.back()[1]) {
                    ans.push_back(intervals[i]);
                } else {
                    ans.back()[1] = max(ans.back()[1], intervals[i][1]);
                }
            }
        }
        return ans;
    }
};
